http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/db/RuntimeServiceDbQueries.scala

```diff
@@ -264,6 +264,236 @@ object RuntimeServiceDbQueries {
+   * @param ownerGoogleProjectIds
+   * @return
+   */
+  private def listAuthorizedRuntimesHelper(
+    // Filters
+    labelMap: LabelMap,
+    excludeStatuses: List[RuntimeStatus],
+    creatorOnly: Option[WorkbenchEmail],
+    workspaceId: Option[WorkspaceId],
+    cloudContextOrCloudProvider: Option[Either[CloudContext, CloudProvider]],
+
+    // Authorizations
+    readerRuntimeIds: List[String],
+    readerWorkspaceIds: List[String],
+    ownerWorkspaceIds: List[String],
+    readerGoogleProjectIds: List[String],
+    ownerGoogleProjectIds: List[String]
+  ): DBIO[Vector[ListRuntimeResponse2]] = {
+
+    // Authorize: show only resources the user is permitted to see
+    val filterVisibleIds = getFilterVisibleIds(readerRuntimeIds,
+                                               readerWorkspaceIds,
+                                               ownerWorkspaceIds,
+                                               readerGoogleProjectIds,
+                                               ownerGoogleProjectIds
+    )
+
+    // Filter: show only user selections
+    val filterWorkspaceId = workspaceId match {
+      case Some(wid) =>
+        List(s"C.`workspaceId` = '${wid.value.toString}'")
+      case None => List.empty
+    }
+
+    val filterCloud = cloudContextOrCloudProvider match {
+      case Some(Left(CloudContext.Gcp(gp))) =>
+        List(s"C.`cloudProvider` = ${CloudProvider.Gcp} AND C.`cloudContext` = '${gp.value}'")
+      case Some(Left(CloudContext.Azure(actx))) =>
+        List(s"C.`cloudProvider` = ${CloudProvider.Azure} AND C.`cloudContext` = '${actx.asString}'")
+      case Some(Right(cloudProvider)) => List(s"C.`cloudProvider` = '${cloudProvider.asString}'")
+      case None                       => List.empty
+    }
+
+    val filterNotExcludedStatus = excludeStatuses.map(s => s"C.`status` != '${s.toString}'")
+
+    val filterCreator = creatorOnly match {
+      case Some(creator) => List(s"C.`creator` = '${creator.value}'")
+      case None          => List.empty
+    }
+
+    val filterClusters = (
+      filterVisibleIds ++
+        filterWorkspaceId ++
+        filterCloud ++
+        filterNotExcludedStatus ++ filterCreator
+    ).filterNot(_.isEmpty).mkString(") AND (")
+
+    val whereFilterClusters = if (filterClusters.isEmpty) "" else s"where (${filterClusters})"
+
+    val whereFilterLabels =
+      if (labelMap.isEmpty)
+        ""
+      else {
+        val query = labelMap
+          .map { case (k, v) =>
+            s"(LABEL.key = '${k}' and LABEL.value = '${v}')"
+          }
+          .mkString(" or ")
+
+        s"""where (
+           |   select
+           |     count(1)
+           |   from
+           |     `LABEL`
+           |   where
+           |     (`resourceId` = FILTERED_CLUSTER.id)
+           |     AND (`resourceType` = 'runtime')
+           |     AND (${query})
+           |    ) = ${labelMap.size}""".stripMargin
+      }
+
+    // Create full query
+    val sqlStatement =
```

Comment:
I know we use raw SQL statements like this a few times in the file, is that necessary? I know the slick syntax gets funky with lots of joins, but I'm a little wary of hardcoding it.

Reply:
I generally support less raw query, more ORM/DAO. This is written this way as a minimal change to the existing code; turning the whole query into slick or something else is a refactor that I hadn't planned for

Reply:
Created https://broadworkbench.atlassian.net/browse/IA-4610

Reply:
And https://broadworkbench.atlassian.net/browse/IA-4626